pkg/mcs/resource_manager/server/token_bukets.go

```diff
@@ -45,100 +52,229 @@ type GroupTokenBucket struct {
 	GroupTokenBucketState `json:"state,omitempty"`
 }
 
+func (gtb *GroupTokenBucket) setState(state *GroupTokenBucketState) {
+	gtb.mu.Lock()
+	defer gtb.mu.Unlock()
+
+	gtb.Tokens = state.Tokens
+	gtb.LastUpdate = state.LastUpdate
+	gtb.Initialized = state.Initialized
+	log.Info("update group token bucket state", zap.Any("state", state))
+}
+
+// TokenSlot is used to split a token bucket into multiple slots to
+// server different clients within the same resource group.
+type TokenSlot struct {
+	// settings is the token limit settings for the slot.
+	settings *rmpb.TokenLimitSettings
+	// assignTokens is the number of tokens in the slot.
+	assignTokens     float64
+	lastAssignTokens float64
+	requireTokensSum float64
+	lastUpdate       *time.Time
+}
+
 // GroupTokenBucketState is the running state of TokenBucket.
 type GroupTokenBucketState struct {
-	Tokens      float64    `json:"tokens,omitempty"`
+	mu     sync.Mutex
+	Tokens float64 `json:"tokens,omitempty"`
+	// ClientUniqueID -> TokenSlot
+	tokenSlots                 map[uint64]*TokenSlot
+	clientConsumptionTokensSum float64
+
 	LastUpdate  *time.Time `json:"last_update,omitempty"`
 	Initialized bool       `json:"initialized"`
 	// settingChanged is used to avoid that the number of tokens returned is jitter because of changing fill rate.
 	settingChanged bool
 }
 
 // Clone returns the copy of GroupTokenBucketState
-func (s *GroupTokenBucketState) Clone() *GroupTokenBucketState {
+func (gts *GroupTokenBucketState) Clone() *GroupTokenBucketState {
+	tokenSlots := make(map[uint64]*TokenSlot)
+	for id, tokens := range gts.tokenSlots {
+		tokenSlots[id] = tokens
+	}
+	var lastUpdate *time.Time
+	if gts.LastUpdate != nil {
+		newLastUpdate := *gts.LastUpdate
+		lastUpdate = &newLastUpdate
+	}
 	return &GroupTokenBucketState{
-		Tokens:      s.Tokens,
-		LastUpdate:  s.LastUpdate,
-		Initialized: s.Initialized,
+		Tokens:                     gts.Tokens,
+		LastUpdate:                 lastUpdate,
+		Initialized:                gts.Initialized,
+		tokenSlots:                 tokenSlots,
+		clientConsumptionTokensSum: gts.clientConsumptionTokensSum,
+	}
+}
+
+func (gts *GroupTokenBucketState) cleanupAssignTokenSum() {
+	gts.clientConsumptionTokensSum = 0
+	for _, slot := range gts.tokenSlots {
+		slot.requireTokensSum = 0
+	}
+}
+
+func (gts *GroupTokenBucketState) balanceSlotTokens(
```

Comment:
Suggest using `XXXLocked()` to indicate the lock should be fetched before the function is called. This is a conventional usage, e.g, https://cs.opensource.google/go/go/+/refs/tags/go1.20.2:src/sync/map.go;l=189.

```suggestion
func (gts *GroupTokenBucketState) balanceSlotTokensLocked(
```